x/distribution/keeper/key.go

```diff
@@ -6,11 +6,13 @@ import (
 
 // keys/key-prefixes
 var (
-	FeePoolKey               = []byte{0x00} // key for global distribution state
-	ValidatorDistInfoKey     = []byte{0x01} // prefix for each key to a validator distribution
-	DelegationDistInfoKey    = []byte{0x02} // prefix for each key to a delegation distribution
-	DelegatorWithdrawInfoKey = []byte{0x03} // prefix for each key to a delegator withdraw info
-	ProposerKey              = []byte{0x04} // key for storing the proposer operator address
+	FeePoolKey                 = []byte{0x00} // key for global distribution state
+	ValidatorDistInfoKey       = []byte{0x01} // prefix for each key to a validator distribution
+	DelegationDistInfoKey      = []byte{0x02} // prefix for each key to a delegation distribution
+	DelegatorWithdrawInfoKey   = []byte{0x03} // prefix for each key to a delegator withdraw info
+	ProposerKey                = []byte{0x04} // key for storing the proposer operator address
+	PreviousTotalPowerKey      = []byte{0x05} // key for storing the total tendermint power on the previous block
+	PreviousValidatorPowersKey = []byte{0x06} // key for storing all validators' tendermint powers on the previous block
```

Comment:
Do we really need these new keys?  Doesn't it already exist as GetBondedValidatorIndexKey?  (e.g. as in stake.Keeper.retrieveLastValidatorSet())

Reply:
GetBondedValidatorIndexKey()/ValidatorsBondedIndexKey (name confusion) is only used in ApplyAndReturnValidatorSetUpdates which gets run at stake.EndBlocker.  Unlike ValidatorsByPowerIndexKey, it only gets updated once at the end of a block.  getValidatorPowerRank isn't related, it's using the other index `ValidatorsByPowerIndexKey`. 

Retrieving the power bytes from the key is implemented in retrieveLastValidatorSet.